codebuild/spec/buildspec_valgrind.yml

```diff
@@ -0,0 +1,77 @@
+---
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use
+# this file except in compliance with the License. A copy of the License is
+# located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing permissions and
+# limitations under the License.
+version: 0.2
+
+batch:
+  build-list:
+    - identifier: gcc_awslc
+      env:
+        compute-type: BUILD_GENERAL1_LARGE
+        image: 024603541914.dkr.ecr.us-west-2.amazonaws.com/docker:ubuntu22codebuild
+        variables:
+          S2N_LIBCRYPTO: awslc
+          COMPILER: gcc
+    - identifier: gcc_openssl_3_0
+      env:
+        compute-type: BUILD_GENERAL1_LARGE
+        image: 024603541914.dkr.ecr.us-west-2.amazonaws.com/docker:ubuntu22codebuild
+        variables:
+          S2N_LIBCRYPTO: openssl-3.0
+          COMPILER: gcc
+    - identifier: gcc_openssl_1_1_1
+      env:
+        compute-type: BUILD_GENERAL1_LARGE
+        image: 024603541914.dkr.ecr.us-west-2.amazonaws.com/docker:ubuntu22codebuild
+        variables:
+          S2N_LIBCRYPTO: openssl-1.1.1
+          COMPILER: gcc
+    - identifier: gcc_openssl_1_0_2
+      env:
+        compute-type: BUILD_GENERAL1_LARGE
+        image: 024603541914.dkr.ecr.us-west-2.amazonaws.com/docker:ubuntu22codebuild
+        variables:
+          S2N_LIBCRYPTO: openssl-1.0.2
+          COMPILER: gcc
+
+phases:
+  pre_build:
+    commands:
+      - |
+        if [ -d "third-party-src" ]; then
+          cd third-party-src;
+        fi
+      - /usr/bin/$COMPILER --version
+  build:
+    on-failure: ABORT
+    commands:
+      - |
+        cmake . -Bbuild \
+          -DCMAKE_C_COMPILER=/usr/bin/$COMPILER \
+          -DCMAKE_PREFIX_PATH=/usr/local/$S2N_LIBCRYPTO \
+          -DCMAKE_BUILD_TYPE=RelWithDebInfo
+      - cmake --build ./build -- -j $(nproc)
+  post_build:
+    on-failure: ABORT
+    commands:
+      - |
+        S2N_VALGRIND=1 \
+        CTEST_PARALLEL_LEVEL=$(nproc) \
+        CTEST_OUTPUT_ON_FAILURE=1 \
+        ctest -T memcheck \
+              --test-dir build;
+              EXITCODE=$?
+      - cd codebuild/bin/
+      - ./display_memory_errors.sh
+      - cd ../../
```

Comment:
This seems new and unexpected. What is the output of ctest without this, and what is the content of the files? Would it make more sense to upload the files to S3?

And why use a whole separate script for one loop? Couldn't you just put the loop in this spec?

Reply:
The reason I add this script is because CTest memcheck automatically add an extra layer of output redirection for Valgrind result. CTest memcheck itself won't display Valgrind error messages in the terminal, instead it will redirect those messages into some `MemoryChecker.*.log` files. In AWS codebuild, users can't directly access those log files, so I wrote a short script to display them after the CTest memecheck run. The script itself can be found in `file changed` section of this PR.

Reply:
I don't think we should upload every memory check's log files into a S3 bucket, but I can move that script into the spec. I think displaying the message on terminal will make it easier to read.

Reply:
Can you give a concrete example of the output of a memory leak from ctest memcheck vs the log file?

Reply:
Take `s2n_stacktrace_test` as an example. Without Valgrind suppression, the current Makefile will run valgrind, and the memory issue caused by the backtrace function will be displayed in terminal (fd=2) like this

```
Running s2n_post_handshake_test.c                          ... Running s2n_tls13_record_aead_test.c                       ... PASSED        557 tests
Running s2n_aead_chacha20_poly1305_test.c                  ... Running s2n_client_key_exchange_test.c                     ... ==10154== 2,304 bytes in 1 blocks are possibly lost in loss record 7 of 7
==10154==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==10154==    by 0x4005D97: malloc (rtld-malloc.h:56)
==10154==    by 0x4005D97: _dlfo_mappings_segment_allocate (dl-find_object.c:217)
==10154==    by 0x4005D97: _dl_find_object_update_1 (dl-find_object.c:671)
==10154==    by 0x4005D97: _dl_find_object_update (dl-find_object.c:804)
==10154==    by 0x400ECCF: dl_open_worker_begin (dl-open.c:735)
==10154==    by 0x4E1CA97: _dl_catch_exception (dl-error-skeleton.c:208)
==10154==    by 0x400DF99: dl_open_worker (dl-open.c:782)
==10154==    by 0x4E1CA97: _dl_catch_exception (dl-error-skeleton.c:208)
==10154==    by 0x400E34D: _dl_open (dl-open.c:883)
==10154==    by 0x4E1CE00: do_dlopen (dl-libc.c:95)
==10154==    by 0x4E1CA97: _dl_catch_exception (dl-error-skeleton.c:208)
==10154==    by 0x4E1CB62: _dl_catch_error (dl-error-skeleton.c:227)
==10154==    by 0x4E1CF36: dlerror_run (dl-libc.c:45)
==10154==    by 0x4E1CF36: __libc_dlopen_mode (dl-libc.c:162)
==10154==    by 0x4DCD610: __libc_unwind_link_get (unwind-link.c:50)
==10154==    by 0x4DCD610: __libc_unwind_link_get (unwind-link.c:40)
==10154==    by 0x4DDBB76: backtrace (backtrace.c:69)
==10154==    by 0x17C50B: s2n_calculate_stacktrace (s2n_errno.c:457)
==10154==    by 0x1235DB: raises_error (s2n_stacktrace_test.c:22)
```

With CTest memcheck, that information will not be displayed on terminal, and the user (if they run build locally) needs to go to `s2n-tls/build/Testing/Temporary/MemoryChecker.230.log` to access that information. On Codebuild, that file will not be accessible for user to access. So I write a script to go through those `MemoryCheker.*.log` and print out error message in codebuild.

Reply:
That is an example of output from the current valgrind makefile job. Can you give a concrete example of the output of a memory leak from new ctest memcheck vs the new log file?

Reply:
I have decided not to use this script, and add `log-fd=2` to valgrind option. Hence, all memory errors will be directly displayed in `stderr`. Here is an [example](https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/Valgrind/build/Valgrind%3Ad9f2d521-9221-44d6-9f87-01acd136a147?region=us-west-2) of the test output.